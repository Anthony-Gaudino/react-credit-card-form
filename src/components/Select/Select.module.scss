@use "../../theme.scss";

.select {
  @include theme.field;
}
