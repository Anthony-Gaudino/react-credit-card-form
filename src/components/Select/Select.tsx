import clsx from "clsx";
import { FormEventHandler } from "react";

import InputLabel from "../InputLabel";
import styles from "./Select.module.scss";

interface SelectProps {
  /**
   * Input field label.
   */
  label: string;
  /**
   * If `true` hides the label. Defaults to false;
   */
  hideLabel?: boolean;
  /**
   * Input field value.
   */
  value?: string;
  /**
   * Input field max string length.
   */
  required?: boolean;
  /**
   * If `true` highlits the input if it's invalid.
   */
  highlightInvalid?: boolean;
  /**
   * Select options.
   */
  children: React.ReactNode;
  /**
   * onChange function callback.
   */
  onChange?: FormEventHandler;
  /**
   * onFocus function callback.
   */
  onFocus?: FormEventHandler;
}

/**
 * A HTML select with a label.
 */
export const Select = (props: SelectProps): JSX.Element => {
  const {
    label,
    children,
    hideLabel = false,
    highlightInvalid = false,
    ...others
  } = props;

  const classes = clsx({
    [styles.select]: true,
    [styles["select--highlight-invalid"]]: highlightInvalid,
  });

  return (
    <InputLabel label={label} hidden={hideLabel}>
      <select className={classes} {...others}>
        {children}
      </select>
    </InputLabel>
  );
};
