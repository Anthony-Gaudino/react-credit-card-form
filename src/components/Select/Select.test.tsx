import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Select from ".";

describe("Test Select component", () => {
  test("the select input is correctly rendered", () => {
    const { asFragment } = render(
      <Select label="My select">
        <option value="" disabled>
          Pick one
        </option>
        <option value="1">Option one</option>
        <option value="2">Option two</option>
      </Select>
    );

    const select = screen.getByText("Pick one");
    const option1 = screen.getByText("Option one");

    expect(select).toBeInTheDocument();
    expect(option1).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("handles click events", async () => {
    render(
      <Select label="My select">
        <option value="" disabled>
          Pick one
        </option>
        <option value="1">Option one</option>
        <option value="2">Option two</option>
      </Select>
    );

    userEvent.selectOptions(
      screen.getByRole("combobox"),
      screen.getByRole("option", { name: "Option one" })
    );

    const selectedOption = screen.getByRole("option", {
      name: "Option one",
    }) as HTMLOptionElement;

    expect(selectedOption.selected).toBe(true);
  });
});
