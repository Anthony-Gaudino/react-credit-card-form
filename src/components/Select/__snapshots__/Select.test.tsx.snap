// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Select component the select input is correctly rendered 1`] = `
<DocumentFragment>
  <label
    class="inputLabel"
  >
    My select
    <select
      class="select"
    >
      <option
        disabled=""
        value=""
      >
        Pick one
      </option>
      <option
        value="1"
      >
        Option one
      </option>
      <option
        value="2"
      >
        Option two
      </option>
    </select>
  </label>
</DocumentFragment>
`;
